src/main/java/org/primeframework/mvc/util/QueryStringBuilder.java

```diff
@@ -49,51 +49,54 @@ public static QueryStringBuilder builder(String uri) {
   }
 
   public QueryStringBuilder beginFragment() {
-    sb.append("#");
-    addSeparator = false;
+    // If query string contains no terms, remove the question mark
+    if (sb.toString().endsWith("?")) {
+      sb.setLength(sb.length() - 1);
+    }
+
+    if (sb.indexOf("#") == -1) {
+      sb.append("#");
+      addSeparator = false;
+    } else {
+      char lastChar = sb.charAt(sb.length() - 1);
+      addSeparator = lastChar != '#' && lastChar != '&';
+    }
```

Comment:
I don't know that this is correct. I believe you can add a `#` after a `?` in a URL. 

For example, I believe this is a valid URL: `https://example.com?foo=bar#section-1`

Reply:
Per our conversation, I think I misunderstood what this was doing.

Reply:
`https://example.com?foo=bar#section-1` is definitely valid.  I was just trying to avoid
`https://example.com?#section-1` but I can certainly allow it.